agent/provider/src/payments/payments.rs

```diff
@@ -212,63 +215,44 @@ async fn send_debit_note(
     Ok(debit_note)
 }
 
-async fn send_invoice(
-    provider_context: &Arc<ProviderCtx>,
-    agreement_id: &str,
-    cost_summary: &CostInfo,
-    activities: &Vec<String>,
-) -> Result<Invoice> {
-    log::info!(
-        "Final cost for agreement [{}]: {}, usage {:?}.",
-        agreement_id,
-        &cost_summary.cost,
-        &cost_summary.usage
-    );
-
-    let invoice = NewInvoice {
-        agreement_id: agreement_id.to_string(),
-        activity_ids: Some(activities.clone()),
-        amount: cost_summary.clone().cost,
-        // TODO: Change this date to meaningful value.
-        //  Now all our invoices are immediately outdated.
-        payment_due_date: Utc::now(),
-    };
-
-    log::debug!("Creating invoice {}.", serde_json::to_string(&invoice)?);
-
-    let payment_api = &provider_context.payment_api;
-    let invoice = payment_api.issue_invoice(&invoice).await.map_err(|error| {
-        anyhow!(
-            "Failed to issue debit note for agreement [{}]. {}",
-            &agreement_id,
-            error
-        )
-    })?;
-
-    log::debug!(
-        "Sending invoice [{}] for agreement [{}].",
-        &invoice.invoice_id,
-        &agreement_id
-    );
-    payment_api
-        .send_invoice(&invoice.invoice_id)
-        .await
-        .map_err(|error| {
-            anyhow!(
-                "Failed to send invoice [{}] for agreement [{}]. {}",
-                &invoice.invoice_id,
-                &agreement_id,
-                error
-            )
-        })?;
-
-    log::info!(
-        "Invoice [{}] sent for agreement [{}].",
-        &invoice.invoice_id,
-        &agreement_id
-    );
+async fn check_invoice_events(provider_ctx: Arc<ProviderCtx>, payments_addr: Addr<Payments>) -> () {
+    let timeout = Some(Duration::from_secs(60));
+    let mut lather_than = Utc::now();
+
+    loop {
+        let events = match provider_ctx
+            .payment_api
+            .get_invoice_events(Some(&lather_than), timeout)
+            .await
+        {
+            Ok(events) => events,
+            Err(e) => {
+                log::error!("Can't query invoice events: {}", e);
+                vec![]
+            }
+        };
```

Comment:
I just saw some new things:
- If you get valid response from `get_invoice_events`, than this call waits until timeout when there are no events in queue, so everything works without active waiting.
    - It would be nice to get this timeout from provider_ctx, instead of using hardcoded value
- In case of error you spam payments api with requests, because get_invoice_events` returns immediately

Reply:
Added a timeout for errors. Both are in `provider_ctx`.